libsolidity/codegen/CompilerUtils.cpp

```diff
@@ -843,6 +848,8 @@ unsigned CompilerUtils::prepareMemoryStore(Type const& _type, bool _padToWordBou
 	else
 	{
 		solAssert(numBytes <= 32, "Memory store of more than 32 bytes requested.");
+		if (_type.category() == Type::Category::Bool)
```

Comment:
In general, this should rather be part of `cleanHigherOrderBits` - please check if that function is called before every memory store.